### PR TITLE
Notification

### PR DESCRIPTION
**Pull request by simonkorz**

### CODE REVIEW

**Comment by LeanderK** [comment]
whohoo!

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java [hunk 1]
```diff
@@ -50,43 +50,61 @@ public NotificationOperations(DatabaseManager manager, String username, String p
...
+     *
+     * @param notificationRecord the NotificationRecord to add the emails to
+     * @return a Notification
      */
-    public NotificationRecord insertNotification(NotificationRecord record) {
-        return create.insertInto(Tables.NOTIFICATION)
-                .set(record)
-                .returning()
-                .fetchOne();
+    private Notification fetchAndAddEmailsToNotification(NotificationRecord notificationRecord) {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java]: i mean thats an optimisation, but there is also a mapList operation defined in getNextRange, so you could alter the signature to `fetchAndAddEmailsToNotifications(List<NotificationRecord> notificationRecords)` and the implementation to `create.....where(notification.id.in(notificationRecords)).map(NotificationTransformer::toProto)`. I don't think it is necessary, but when you have 50 notifications defined it results in 50+ db-calls.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java [hunk 2]
```diff
@@ -122,60 +137,80 @@ public boolean delteNotification(int notificationID) {
      * @return {@code true} if deleted, {@code false} otherwise
      */
     public boolean deleteNotification(int id) {
-        NotificationRecord record = create.newRecord(NOTIFICATION);
-        record.setIdNotification(id);
-
-        return create.executeDelete(record, NOTIFICATION.ID_NOTIFICATION.eq(id)) == 1;
+        create.deleteFrom(NOTIFICATION_RECEIVER_EMAIL)
+                .where(NOTIFICATION_RECEIVER_EMAIL.NOTIFICATION.eq(id));
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java]: they delete themselves if the notification gets deleted. same with Notification_token

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java (hunk 1 above)
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java]: also the method-name is not the best. it does not modify the database, fetchNotification()? if the type is the proto it explains itself

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java [hunk 3]
```diff
@@ -122,60 +137,80 @@ public boolean delteNotification(int notificationID) {
...
 
-        record = NotificationTransformer.mergeRecord(record, notification);
-        record.update();
+        notificationRecord = NotificationTransformer.mergeRecord(notificationRecord, notification);
+        notificationRecord.update();
+
+        notification.getEmailsList().stream()
+                .map(s -> new NotificationReceiverEmailRecord(null, notification.getId(), s))
+                .collect(Collectors.collectingAndThen(Collectors.toList(), create::batchUpdate))
+                .execute();
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java]: this won't work. update needs a primary key and you never delete existing. (i would do the following: first transform then in NotificationReceiverEmailRecords. then create a **transaction** that deletes the associated and inserts new ones)

Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java]: Just use the implementation from #83.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java [hunk 4]
```diff
@@ -122,60 +137,80 @@ public boolean delteNotification(int notificationID) {
...
+     * @param newTokenRecords the new records from a current fetch query
+     * @param notificationId  the id of the notification
+     * @return a list of newly added NotificationTokenRecords
+     */
+    public List<NotificationTokenRecord> diffTokenRecords(Map<Integer, NotificationTokenRecord> newTokenRecords, int notificationId) {
+        List<Query> queries = new ArrayList<>();
+
+        try (Cursor<NotificationTokenRecord> recordCursor = create.selectFrom(NOTIFICATION_TOKEN)
+                .where(NOTIFICATION_TOKEN.NOTIFICATION.eq(notificationId))
+                .fetchLazy()) {
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java]: .stream() would be possible to, you still need to use try with resources though

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/Notification.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    private static final Logger LOGGER = LogManager.getLogger(Notification.class);
+
+    private int id;
+    private String name;
+    private String description;
+    private int checkPeriod;
+    private String query;
+    private boolean sendOnce = false;
+    private List<String> receiverEmails;
+    private NotificationPolicy policy;
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/Notification.java]: i would add some finals

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/NotificationController.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+     */
+    public void init() {
+        loadNotificationsFromDatabase();
+    }
+
+    private void loadNotificationsFromDatabase() {
+        List<edu.kit.ipd.crowdcontrol.objectservice.proto.Notification> notificationList = operations.getAllNotifications();
+        for (edu.kit.ipd.crowdcontrol.objectservice.proto.Notification notificationProto : notificationList) {
+            createNotification(Notification.fromProtobuf(notificationProto, policy));
+        }
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/NotificationController.java]: using operations.getAllNotifications().foreach would be more beautiful

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * The SQLEmailNotificationPolicy checks given queries in a SQL-Database and sends notifications via email.
+ *
+ * @author Simon Korz
+ * @version 1.0
+ */
+public class SQLEmailNotificationPolicy extends NotificationPolicy<List<String>> {
+    private MailSender mailSender;
+    private NotificationOperations operations;
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java]: final?

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    private boolean resultHasIdAndTokenField(Result<Record> result) {
+        Field<Integer> idField = DSL.field("id", Integer.class);
+        Field<String> tokenField = DSL.field("token", String.class);
+
+        return (result.get(0).field(idField) != null) && (result.get(0).field(tokenField) != null);
+    }
+
+    @Override
+    protected List<String> check(Notification notification) {
+        Result<Record> result = operations.runReadOnlySQL(notification.getQuery());
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java]: i would turn this lazy too, i can give you a Stream<X> and you have to use it in a try with resources block. this operation should be more expensive than operations.diffTokenRecords

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                message.append(Template.apply(notification.getDescription(), placeholderMap));
+            } catch (IllegalArgumentException iae) {
+                message.append(notification.getDescription());
+            }
+        }
+
+        String subject = "[CrowdControl] " + notification.getName();
+        try {
+            for (String receiver : notification.getReceiverEmails())
+            mailSender.sendMail(receiver, subject, message.toString());
```
Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java]: we are in a curly braces land..., i would also catch the exception for each sendMail operation.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java (hunk 3 above)
Comment (simonkorz) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/database/operations/NotificationOperations.java]: Okay I didn't know that, I think I copied it from another operations file.

**Comment by LeanderK** [comment]
do you validate the email in the resource-class? So that i can't submit `_!A` as an notification receiver email.

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/NotificationController.java (hunk 6 above)
Comment (simonkorz) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/NotificationController.java]: thats arguable ;)

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java (hunk 8 above)
Comment (simonkorz) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java]: yep I didn't yet think of optimization too much, but thats a good idea

Comment (LeanderK) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java]: its not a deal breaker....i just wondered why you optimised the cheaper method ;)

File: src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java (hunk 9 above)
Comment (kelunik) [inline src/main/java/edu/kit/ipd/crowdcontrol/objectservice/notification/SQLEmailNotificationPolicy.java]: Curly Braces Land :tada:

**Comment by kelunik** [comment]
I'll do that in my PR.

**Pull request closed without merging**